File: github-dashboard/src/aggregators.ts
import type {
  GitHubContributor,
  GitHubCommitDetail,
  GitHubPullRequest,
  ContributorStat,
  FileStat,
  CommitGranularityStat,
  BranchMergeStat,
} from './types'

// ── Use Case 1: Commits by user ──────────────────────────────────────────────

export function aggregateContributors(
  raw: GitHubContributor[],
): ContributorStat[] {
  return raw
    .filter((c) => c.author !== null)
    .map((c) => ({
      login: c.author!.login,
      avatarUrl: c.author!.avatar_url,
      totalCommits: c.total,
    }))
    .sort((a, b) => b.totalCommits - a.totalCommits)
}

// ── Use Case 2: Most-changed files ──────────────────────────────────────────

export function aggregateFileChurn(details: GitHubCommitDetail[]): FileStat[] {
  const counts = new Map<string, number>()
  for (const commit of details) {
    for (const file of commit.files ?? []) {
      counts.set(file.filename, (counts.get(file.filename) ?? 0) + 1)
    }
  }
  return Array.from(counts.entries())
    .map(([filename, changeCount]) => ({ filename, changeCount }))
    .sort((a, b) => b.changeCount - a.changeCount)
}

// ── Use Case 3: Commit granularity ──────────────────────────────────────────

export function aggregateGranularity(
  details: GitHubCommitDetail[],
): CommitGranularityStat[] {
  return details.map((c) => ({
    sha: c.sha,
    filesChanged: c.files?.length ?? 0,
    authorLogin: c.author?.login ?? c.commit.author?.name ?? 'unknown',
    date: c.commit.author?.date ?? '',
    message: c.commit.message.split('\n')[0].slice(0, 72),
  }))
}

// ── Use Case 4: Branch-to-merge duration ────────────────────────────────────

/**
 * For each merged PR we have `created_at` (when the PR was opened) and
 * `merged_at`. We use PR created_at as a reliable proxy for branch activity
 * start because the per-commit detail calls needed for the first-commit-on-branch
 * approach would consume the entire unauthenticated rate budget.
 *
 * If a GitHub token is supplied upstream, callers can enrich this with the
 * actual first commit date via fetchCommitsForBranch.
 */
export function aggregateBranchMerge(
  prs: GitHubPullRequest[],
): BranchMergeStat[] {
  return prs
    .filter((pr) => pr.merged_at !== null)
    .map((pr) => {
      const start = new Date(pr.created_at)
      const end = new Date(pr.merged_at!)
      const durationHours = Math.max(
        0,
        (end.getTime() - start.getTime()) / (1000 * 60 * 60),
      )
      return {
        prNumber: pr.number,
        title: pr.title,
        branchName: pr.head.ref,
        authorLogin: pr.user?.login ?? 'unknown',
        firstCommitDate: pr.created_at,
        mergedAt: pr.merged_at!,
        durationHours,
      }
    })
    .sort((a, b) => new Date(b.mergedAt).getTime() - new Date(a.mergedAt).getTime())
}

File: GithubDashboard.Api/Aggregators.cs
namespace GithubDashboard;

public static class Aggregators
{
    // ── UC1: Commits by contributor ───────────────────────────────────────────

    public static List<ContributorStat> AggregateContributors(IEnumerable<GhContributor> raw) =>
        raw.Where(c => c.Author is not null)
           .Select(c => new ContributorStat(c.Author!.Login, c.Author.AvatarUrl, c.Total))
           .OrderByDescending(c => c.TotalCommits)
           .ToList();

    // ── UC2: Most-changed files ───────────────────────────────────────────────

    public static List<FileStat> AggregateFileChurn(IEnumerable<GhCommitDetail> details)
    {
        var counts = new Dictionary<string, int>(StringComparer.Ordinal);
        foreach (var commit in details)
            foreach (var file in commit.Files ?? [])
                counts[file.Filename] = counts.GetValueOrDefault(file.Filename) + 1;

        return counts
            .Select(kv => new FileStat(kv.Key, kv.Value))
            .OrderByDescending(f => f.ChangeCount)
            .ToList();
    }

    // ── UC3: Commit granularity ───────────────────────────────────────────────

    public static List<CommitGranularityStat> AggregateGranularity(IEnumerable<GhCommitDetail> details) =>
        details.Select(c => new CommitGranularityStat(
            c.Sha,
            c.Files?.Length ?? 0,
            c.Author?.Login ?? c.Commit.Author?.Name ?? "unknown",
            c.Commit.Author?.Date ?? string.Empty,
            FirstLine(c.Commit.Message, 72)))
        .ToList();

    // ── UC4: Branch-to-merge duration ─────────────────────────────────────────

    public static List<BranchMergeStat> AggregateBranchMerge(IEnumerable<GhPr> prs) =>
        prs.Where(pr => pr.MergedAt is not null)
           .Select(pr =>
           {
               var start = DateTime.Parse(pr.CreatedAt, null, System.Globalization.DateTimeStyles.RoundtripKind);
               var end   = DateTime.Parse(pr.MergedAt!, null, System.Globalization.DateTimeStyles.RoundtripKind);
               var hours = Math.Max(0, (end - start).TotalHours);
               return new BranchMergeStat(
                   pr.Number,
                   pr.Title,
                   pr.Head.Ref,
                   pr.User?.Login ?? "unknown",
                   pr.CreatedAt,
                   pr.MergedAt!,
                   Math.Round(hours, 2));
           })
           .OrderByDescending(s => DateTime.Parse(s.MergedAt, null, System.Globalization.DateTimeStyles.RoundtripKind))
           .ToList();

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static string FirstLine(string msg, int maxLen)
    {
        var line = msg.Split('\n')[0];
        return line.Length <= maxLen ? line : line[..maxLen];
    }
}
